django/contrib/admin/options.py

```diff
@@ -1305,6 +1306,16 @@ def add_view(self, request, form_url='', extra_context=None):
         ModelForm = self.get_form(request)
         if request.method == 'POST':
             form = ModelForm(request.POST, request.FILES)
+
+            # if original is set, set initial file fields
+            if original:
+                for field_name, field in form.fields.items():
```

Comment:
It doesn't seem like a proper separation of concerns to have `isinstance()` checks for specific field types here and in `ModelForm` methods. Unfortunately, I can't make an alternate recommendation without really digging into this.